File: frontend/src/api/customers.ts
import api from "./http";

export type CustomerRow = {
  id: string;
  name: string;
  manager: string;
  node: string;
  node_id: number;
  remote_email: string;
  sub_id?: string;
  renew_price: string;
  expiry_date: string;
  duration: string;
  traffic?: string;
  traffic_total_gb?: number;
  traffic_used_gb?: number;
  traffic_remaining_gb?: number | null;
  traffic_total_display?: string;
  traffic_used_display?: string;
  traffic_remaining_display?: string;
  is_unlimited_traffic?: boolean;
  webhook_url?: string;
  enable?: boolean;
  expiry_display?: string;
  is_unlimited_expiry?: boolean;
  inbound_ids?: number[];
  total_gb?: number;
  traffic_multiplier?: number;
  limit_ip?: number;
  inbounds?: Array<{ id: number; remark: string; protocol: string; port: number }>;
  remaining_days: number;
  status_text: string;
  status_level: "expired" | "today" | "warning" | "healthy" | "disabled" | "unlimited";
};

export type CustomerSubscription = {
  customer_id: string;
  name: string;
  node: string;
  sub_id: string;
  links: {
    standard?: string;
    json?: string;
    clash?: string;
  };
  remote_links?: {
    standard?: string;
    json?: string;
    clash?: string;
  };
  local_links?: {
    standard?: string;
    json?: string;
    clash?: string;
  };
  local_subscription?: {
    enabled: boolean;
    base_url: string;
    port: number;
    title: string;
  };
  protocol_links: string[];
};

export type CustomerListResponse = {
  count: number;
  data: CustomerRow[];
};

export type CustomerAuditRow = {
  id: number;
  customer_name: string;
  action: string;
  actor: string;
  change_summary: string;
  created_at: string;
};

export type CustomerRenewalRow = {
  id: number;
  customer_name: string;
  actor: string;
  renew_days: number;
  old_expiry: string;
  new_expiry: string;
  renew_price: string;
  created_at: string;
};

export type CustomerPayload = {
  name: string;
  manager?: string;
  node: string;
  node_id?: number;
  remote_email?: string;
  expiry_date?: string;
  renew_price?: string;
  webhook_url?: string;
  duration_mode?: "days" | "date";
  duration_days?: number;
  custom_expiry_date?: string;
  inbound_ids?: number[];
  total_gb?: number;
  traffic_multiplier?: number;
  enable?: boolean;
  limit_ip?: number;
};

export type RenewPayload = {
  renew_days: number;
  renew_price?: string;
};

export type BulkUpdateFieldsPayload = {
  customer_ids: string[];
  enable?: boolean;
  total_gb?: number;
  traffic_multiplier?: number;
  limit_ip?: number;
  renew_price?: string;
  duration_mode?: "days" | "date";
  duration_days?: number;
  custom_expiry_date?: string;
};

export async function fetchCustomers(params: { keyword?: string; node?: string; node_id?: number | string; manager?: string } = {}) {
  const response = await api.get("/api/v1/customers", { params });
  return response.data as { count: number; data: CustomerRow[] };
}

export async function fetchCustomerDetail(customerId: string) {
  const response = await api.get(`/api/v1/customers/${customerId}`);
  return response.data.data;
}

export async function createCustomer(payload: CustomerPayload) {
  const response = await api.post("/api/v1/customers", payload);
  return response.data as { success: boolean; message: string; customer_id?: string };
}

export async function updateCustomer(customerId: string, payload: Partial<CustomerPayload>) {
  const response = await api.put(`/api/v1/customers/${customerId}`, payload);
  return response.data as { success: boolean; message: string };
}

export async function deleteCustomer(customerId: string) {
  const response = await api.delete(`/api/v1/customers/${customerId}`);
  return response.data as { success: boolean; message: string };
}

export async function renewCustomer(customerId: string, payload: RenewPayload) {
  const response = await api.post(`/api/v1/customers/${customerId}/renew`, payload);
  return response.data as { success: boolean; message: string; new_expiry: string; renew_price: string };
}

export async function resetCustomerTraffic(customerId: string) {
  const response = await api.post(`/api/v1/customers/${customerId}/reset-traffic`);
  return response.data as { success: boolean; message: string; data?: CustomerRow };
}

export async function bulkAssignCustomerManager(customerIds: string[], manager: string) {
  const response = await api.post("/api/v1/customers/bulk/assign-manager", { customer_ids: customerIds, manager });
  return response.data as { success: boolean; message: string; total: number; updated: unknown[]; errors: Array<{ id: string; message: string }> };
}

export async function bulkUpdateCustomers(payload: BulkUpdateFieldsPayload) {
  const response = await api.post("/api/v1/customers/bulk/update-fields", payload);
  return response.data as { success: boolean; message: string; total: number; updated: unknown[]; errors: Array<{ id: string; message: string }> };
}

export async function fetchCustomerSubscription(customerId: string) {
  const response = await api.get(`/api/v1/customers/${customerId}/subscription`);
  return response.data.data as CustomerSubscription;
}

export async function fetchCustomerAudit(customerId: string) {
  const response = await api.get(`/api/v1/customers/${customerId}/audit`);
  return response.data.data as CustomerAuditRow[];
}

export async function fetchCustomerAuditByAction(customerId: string, action?: string) {
  const response = await api.get(`/api/v1/customers/${customerId}/audit`, { params: action ? { action } : undefined });
  return response.data.data as CustomerAuditRow[];
}

export async function fetchCustomerRenewals(customerId: string) {
  const response = await api.get(`/api/v1/customers/${customerId}/renewals`);
  return response.data.data as CustomerRenewalRow[];
}

export async function testCustomerWebhook(customerId: string) {
  const response = await api.post(`/api/v1/customers/${customerId}/test-webhook`);
  return response.data as { success: boolean; message: string };
}

File: backend/app/api/v1/customers.py
from fastapi import APIRouter, Depends, Request
from pydantic import BaseModel, Field

from backend.app.core.config import Settings
from backend.app.core.deps import SESSION_ROLE_KEY, require_login
from backend.app.services.dashboard import clear_dashboard_cache
from backend.app.services.customers import (
    bulk_assign_manager,
    bulk_update_customer_fields,
    create_customer,
    customer_audit_logs,
    customer_renewal_logs,
    delete_customer,
    get_customer,
    get_customer_subscription,
    list_customers,
    process_customer_renew,
    reset_customer_traffic,
    update_customer,
)
from backend.app.services.notifications import test_customer_webhook

router = APIRouter(prefix="/api/v1/customers", tags=["customers"])


class CustomerPayload(BaseModel):
    name: str | None = None
    manager: str | None = None
    node: str | None = None
    node_id: int | None = None
    remote_email: str | None = None
    expiry_date: str | None = None
    renew_price: str | None = None
    webhook_url: str | None = None
    duration_mode: str | None = None
    duration_days: int | None = None
    custom_expiry_date: str | None = None
    inbound_ids: list[int] | None = None
    total_gb: float | None = None
    traffic_multiplier: float | None = None
    enable: bool | None = None
    limit_ip: int | None = None


class RenewPayload(BaseModel):
    renew_days: int
    renew_price: str | None = None


class BulkAssignManagerPayload(BaseModel):
    customer_ids: list[str]
    manager: str


class BulkUpdateFieldsPayload(BaseModel):
    customer_ids: list[str]
    enable: bool | None = None
    total_gb: float | None = None
    traffic_multiplier: float | None = None
    limit_ip: int | None = None
    renew_price: str | None = None
    duration_mode: str | None = None
    duration_days: int | None = None
    custom_expiry_date: str | None = None
    expiry_date: str | None = None


class AuditQuery(BaseModel):
    action: str | None = None


@router.get("")
def customers(request: Request, keyword: str = "", node: str = "", node_id: int | None = None, manager: str = "", username: str = Depends(require_login)):
    settings: Settings = request.app.state.settings
    actor_role = str(request.session.get(SESSION_ROLE_KEY) or "user")
    rows = list_customers(settings, keyword=keyword, node_filter=node, node_id_filter=node_id, manager_filter=manager, actor=username, actor_role=actor_role)
    return {"success": True, "code": 0, "msg": "ok", "count": len(rows), "data": rows}


@router.post("")
def create(request: Request, payload: CustomerPayload, username: str = Depends(require_login)):
    settings: Settings = request.app.state.settings
    actor_role = str(request.session.get(SESSION_ROLE_KEY) or "user")
    result = create_customer(settings, payload.model_dump(exclude_none=True), actor=username, actor_role=actor_role)
    clear_dashboard_cache()
    return result


@router.post("/bulk/assign-manager")
def bulk_assign(request: Request, payload: BulkAssignManagerPayload, username: str = Depends(require_login)):
    settings: Settings = request.app.state.settings
    actor_role = str(request.session.get(SESSION_ROLE_KEY) or "user")
    result = bulk_assign_manager(settings, payload.customer_ids, payload.manager, actor=username, actor_role=actor_role)
    clear_dashboard_cache()
    return result


@router.post("/bulk/update-fields")
def bulk_update(request: Request, payload: BulkUpdateFieldsPayload, username: str = Depends(require_login)):
    settings: Settings = request.app.state.settings
    actor_role = str(request.session.get(SESSION_ROLE_KEY) or "user")
    result = bulk_update_customer_fields(
        settings,
        payload.customer_ids,
        payload.model_dump(exclude={"customer_ids"}, exclude_none=True),
        actor=username,
        actor_role=actor_role,
    )
    clear_dashboard_cache()
    return result


@router.post("/{customer_id:path}/renew")
def renew(request: Request, customer_id: str, payload: RenewPayload, username: str = Depends(require_login)):
    settings: Settings = request.app.state.settings
    actor_role = str(request.session.get(SESSION_ROLE_KEY) or "user")
    result = process_customer_renew(settings, customer_id, payload.renew_days, payload.renew_price or "", actor=username, actor_role=actor_role)
    clear_dashboard_cache()
    return result


@router.post("/{customer_id:path}/reset-traffic")
def reset_traffic(request: Request, customer_id: str, username: str = Depends(require_login)):
    settings: Settings = request.app.state.settings
    actor_role = str(request.session.get(SESSION_ROLE_KEY) or "user")
    result = reset_customer_traffic(settings, customer_id, actor=username, actor_role=actor_role)
    clear_dashboard_cache()
    return result


@router.get("/{customer_id:path}/audit")
def audit(request: Request, customer_id: str, action: str | None = None, username: str = Depends(require_login)):
    settings: Settings = request.app.state.settings
    actor_role = str(request.session.get(SESSION_ROLE_KEY) or "user")
    rows = customer_audit_logs(settings, customer_id, actor=username, actor_role=actor_role)
    if action:
        rows = [r for r in rows if r.get("action") == action]
    return {"success": True, "data": rows}


@router.post("/{customer_id:path}/test-webhook")
def webhook_test(request: Request, customer_id: str, username: str = Depends(require_login)):
    settings: Settings = request.app.state.settings
    actor_role = str(request.session.get(SESSION_ROLE_KEY) or "user")
    return test_customer_webhook(settings, customer_id, actor=username, actor_role=actor_role)


@router.get("/{customer_id:path}/subscription")
def subscription(request: Request, customer_id: str, username: str = Depends(require_login)):
    settings: Settings = request.app.state.settings
    actor_role = str(request.session.get(SESSION_ROLE_KEY) or "user")
    return get_customer_subscription(settings, customer_id, actor=username, actor_role=actor_role)


@router.get("/{customer_id:path}/renewals")
def renewals(request: Request, customer_id: str, username: str = Depends(require_login)):
    settings: Settings = request.app.state.settings
    actor_role = str(request.session.get(SESSION_ROLE_KEY) or "user")
    return {"success": True, "data": customer_renewal_logs(settings, customer_id, actor=username, actor_role=actor_role)}


@router.get("/{customer_id:path}")
def detail(request: Request, customer_id: str, username: str = Depends(require_login)):
    settings: Settings = request.app.state.settings
    actor_role = str(request.session.get(SESSION_ROLE_KEY) or "user")
    return {"success": True, "data": get_customer(settings, customer_id, actor=username, actor_role=actor_role)}


@router.put("/{customer_id:path}")
def edit(request: Request, customer_id: str, payload: CustomerPayload, username: str = Depends(require_login)):
    settings: Settings = request.app.state.settings
    actor_role = str(request.session.get(SESSION_ROLE_KEY) or "user")
    result = update_customer(settings, customer_id, payload.model_dump(exclude_none=True), actor=username, actor_role=actor_role)
    clear_dashboard_cache()
    return result


@router.delete("/{customer_id:path}")
def remove(request: Request, customer_id: str, username: str = Depends(require_login)):
    settings: Settings = request.app.state.settings
    actor_role = str(request.session.get(SESSION_ROLE_KEY) or "user")
    result = delete_customer(settings, customer_id, actor=username, actor_role=actor_role)
    clear_dashboard_cache()
    return result
